File: 39.py
class Solution:
    def __init__(self):
        # map dict sum of array to array
        self.ans = []

    def combinationSum(self, candidates: list[int], target: int) -> list[list[int]]:
        self.gen_combs(candidates, target, [], 0)
        return self.ans

    def gen_combs(self, candidates, target, curr, idx):
        if target == 0:
            self.ans.append(curr)
            return
        elif target < 0:
            return
        for i in range(idx, len(candidates)):
            elt = candidates[i]
            self.gen_combs(candidates, target - elt, curr + [elt], i)
        return


print(Solution().combinationSum([2, 3, 5], 8))

File: 39.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
class Solution {
  vector<vector<int>> ans;

public:
  void bt(vector<int> &candidates, vector<int> &curr, int target, int idx) {
    if (target == 0) {
      ans.push_back(curr);
      return;
    } else if (target < 0) {
      return;
    }
    for (int i = idx; i < candidates.size(); ++i) {
      curr.push_back(candidates[i]);
      bt(candidates, curr, target - candidates[i], i);
      curr.pop_back();
    }
  }
  vector<vector<int>> combinationSum(vector<int> &candidates, int target) {
    this->ans = {};
    vector<int> curr = {};
    bt(candidates, curr, target, 0);
    return ans;
  }
};
int main() {}